### ISSUE TITLE
Fetcherror no backend connection

### ISSUE THREAD

**Issue by markhowells** [nigoroll/libvmod-dynamic#4]
I'm trying to create a simple proxy for a site. As the details are unknown until runtime vmod_dynamic looked perfect.  However about 30% of the time i see the following in the log

```
-   VCL_call       BACKEND_FETCH
-   VCL_return     fetch
-   FetchError     no backend connection
```

Here's a tiny VCL that causes the problem.

```
vcl 4.0;
import dynamic;
backend default {
    .host = "10.0.0.1";
    .port = "80";
}
sub vcl_init {
    new dynamic_dir = dynamic.director(  port="80" );
}
sub vcl_recv {
    set req.backend_hint = dynamic_dir.backend();
}
sub vcl_backend_fetch {
    set bereq.backend = dynamic_dir.backend();
}

```

As it stands, I'd expect that to proxy any traffic that reaches it (I'm not expecting this to cache or anything yet). I've been testing using a range of sites but a simple test add the following to your hosts file on a desktop
`<ip of your varnish> www.dalestrailseries.co.uk`
and point your browser at http://www.dalestrailseries.co.uk 

At best you'll either see a whole page or a broken page or just an error, click a sidebar link or two and it'll always break. The same issue exists with every site i've set up...  I've experimented with every timeout I can and nothing seems to change

I should add that I'm totally new to Varnish - I apologise in advance if I've got the wrong end of the stick somewhere...

I've attached the varnishlog

[varnishlog.txt](https://github.com/nigoroll/libvmod-dynamic/files/558154/varnishlog.txt)

**Comment by Dridi** [comment]
Unfortunately, the logs don't tell much. Could you please try again and provide the output for both commands while you were performing tests?

```
varnishlog -g request
varnishlog -g raw | awk '{if($1 = "0") print}'
```

We need to know what's happening outside of HTTP transactions too.

**Comment by markhowells** [comment]
[varnishlog -g request](http://pastebin.com/izRF5G9E)

[varnishlog -g raw | awk '{if($1 = "0") print}'](http://pastebin.com/NYi4rTfs)

**Comment by Dridi** [comment]
Apologies, there was a typo in my `awk` command, it's missing an equal sign.

```
varnishlog -g request
varnishlog -g raw | awk '{if($1 == "0") print}'
```

I should have tried the commands before posting them :-/

**Comment by markhowells** [comment]
And I should have spotted that before I blindly copy/pasted it ... :)

I suspect something's wrong though...

All the second one does is show the PING/PONG lines - no other output

I have discovered that a browser refresh always causes a 503. The first visit to one of the pages always works. So it 'feels' like alternate requests work then fail.

I've left the Varnish instance running (it's only a DigitalOcean droplet so no harm leaving it open while) If you'd like to add 
`139.59.179.7 www.dalestrailseries.co.uk`
to your hosts file and browse to http://www.dalestrailseries.uk/dt20/ and refresh a few times you'll see what I mean

**Comment by markhowells** [comment]
On the off-chance you meant ' != ' ...

[vlograw.txt](https://github.com/nigoroll/libvmod-dynamic/files/558376/vlograw.txt)

**Comment by Dridi** [comment]
Since you aren't running Varnish at scale, we may still find the relevant logs in the past, updated commands:

```
varnishlog -d -g request
varnishlog -d -g raw
```

No need to perform further tests yet.

I can confirm it is `==`, the logs I was expecting are documented here:
https://github.com/nigoroll/libvmod-dynamic/blob/master/src/vmod_dynamic.vcc#L49-L61

However this time I will do the `awk` filtering locally.

**Comment by markhowells** [comment]
[request.txt](https://github.com/nigoroll/libvmod-dynamic/files/558415/request.txt)
[raw.txt](https://github.com/nigoroll/libvmod-dynamic/files/558414/raw.txt)

**Comment by markhowells** [comment]
Hi,
I've been doing some more digging and I'mn wondering whether the following varnishstat means anything. There appear to be two entries for vmod_dynamic

`VBE.boot.dynamic_dir(104.28.18.83).beresp_hdrbytes                   1.53K         0.00          6.00          0.00          0.00          0.00
VBE.boot.dynamic_dir(104.28.18.83).beresp_bodybytes                 27.25K         0.00        123.00          0.00          0.00          0.00
VBE.boot.dynamic_dir(104.28.18.83).conn                                 0          0.00           .            0.00          0.01          0.01
VBE.boot.dynamic_dir(104.28.18.83).req                                  4          0.00           .            0.00          0.01          0.01
VBE.boot.dynamic_dir(104.28.19.83).bereq_hdrbytes                    1.39K         0.00          6.00          0.00          0.00          0.00
VBE.boot.dynamic_dir(104.28.19.83).beresp_hdrbytes                   1.18K         0.00          5.00          0.00          0.00          0.00
VBE.boot.dynamic_dir(104.28.19.83).beresp_bodybytes                  6.08K         0.00         27.00          0.00          0.00          0.00
VBE.boot.dynamic_dir(104.28.19.83).req                                  3          0.00           .            0.00          0.00          0.00`
Like I said, I'm a total novice regarding Varnish, but if you could give me a steer I'm happy to do some more investigation (I can read/write C)

**Comment by markhowells** [comment]
Hi Dridi - Can I ask if you'll be needing the varnish server that demonstrates this issue? It's currently unfirewalled and I don't really want to keep it lying around for too long.

**Comment by Dridi** [comment]
Hi,

Sorry for not being very responsive, both @nigoroll and I are busy with other things. Or at least I'm assuming he is too. I should be able to have quiet look at that this weekend.

Quick answer, your two entries reported by `varnishstat` are different, there's a different IP address (104.28.19.83). Unless you use a probe reporting sick or actively mark the backends as sick via the CLI. Otherwise the only way to hit this error is to have no backend at all in the director.

Either a lookup came up with no results, or it came after the `first_lookup_timeout`. Both hypotheses apply for the first lookup, but according to what you described it's not the case. So until we have some useful debugging in place it's hard to tell, but that's what I'm planning to do first.

Dridi

**Comment by Dridi** [comment]
Hi @markhowells,

The current master branch contains a debug mode you can enable (preferably in `vcl_init`) that could help troubleshoot your connection problem. This time I will ask a different approach:

- start varnish
- run `varnishlog -g raw -w output.vsl`
- perform your tests
- stop varnish
- stop varnishlog
- attach `output.vsl`

There's room for more debugging but I think we can start with that. Could you please give it a spin?

Thanks

**Comment by markhowells** [comment]
OK. Here's the required output.vsl.

Thr master branch was compiled against and run on the varnish-cache master.  I used the same test VCL as above (with the exception of `dynamic_dir.debug(1)` in `vcl_init` ).  I spoofed the IP for www.dalestrailseries.uk and www.dalestrailseries.uk to point at varnish-cache.  Firefox loads the page http://www.dalestrailseries.uk correctly, then upon hitting refresh I see the error as before. 

As the proxy has been around a while it looks like it's recieving a lot of traffic, so there may be some unwanted cruft in the vsl (attached). If you want anything else just shout.

Cheers

[output.zip](https://github.com/nigoroll/libvmod-dynamic/files/590739/output.zip)

**Comment by Dridi** [comment]
Very interesting, I see the problem clearly happening and now I'm starting to suspect it is related to #7 and could be fixed by #8. Could you please build #8 and give it a try?

Thanks

**Comment by markhowells** [comment]
No change.

Using #8 directly I had to comment out the `dynamic.debug(1)` line in vcl_init

So I merged #8 with master and it ran fine - I presume that's what you wanted.  Output.vsl generated with `varnishlog -g raw -q '* ~ dynamic' -w output.vsl`

NB **-w** not **-r** ...

[output.zip](https://github.com/nigoroll/libvmod-dynamic/files/591918/output.zip)

**Comment by Dridi** [comment]
> NB **-w** not **-r** ...

Sorry about that, I had posted the command I used to see the problem, hence the `-r`. Filteriing when reading.

> So I merged #8 with master and it ran fine - I presume that's what you wanted.

Yes, sorry, I wrote that in a hurry.

> No change.

That's very unfortunate because now I'm back to a blank sheet. In the mean time let me suggest a workaround in VCL.

```
sub vcl_backend_error {
    if (bereq.restarts == 0) {
        return (restart);
    }
}
```

Do a single retry if we didn't get a backend connection (assuming it was the error).

**Comment by markhowells** [comment]
The VCL workround didn't work for Varnish v5. I ended up with
`sub vcl_backend_error {
        return (retry);
}`
which does appear to solve the issue - well assuming that it doesn't introduce some infinite loop.

**Comment by Dridi** [comment]
Yes, I mixed up retry and restart :-/

```
sub vcl_backend_error {
    if (bereq.retries == 0) {
        return (retry);
    }
}
```

FYI if you retry unconditionally, it can happen up to `max_retries` time.

```
varnishadm param.show max_retries
max_retries
        Value is: 4 [retries] (default)
        Minimum is: 0

        Upper limit on how many times a backend fetch can retry.
```

No infinite loop even with your snippet.

**Comment by Dridi** [comment]
Dammit, backtracking one more time, I just realized that I was looking at the wrong error.

When skimming through the logs I was looking at a different `FetchError` [1], and this is *why* I thought it could be related to #7 which I'd like to see happen on my machine.

In your case [2] it's a Varnish problem. Basically, Varnish will keep connections open (default behavior in HTTP/1.1) until `backend_idle_timeout` [3] triggers. Your backend may have a lower [4] setting for keep-alive'd sessions, so after a burst of traffic you may accumulate open connections in the sessions pool and later (let's say 30s later) Varnish will try reusing dead connections.

Under the hood, Varnish does one retry at most in the case of a failed connection, so if you get two dead sessions in a row you end up with this `FetchError`. Your only escape is to then do the retry in VCL.

You can also set the timeout to something slightly lower than your backend's timeout for client connections (Varnish being the client here) or increase the backend's keep-alive for better performance. The point is to have matching expectations on both sides of a connection.

Let me know if it makes sense to you. Thanks to my blindness we have a debug mode, but it doesn't look like a bug anymore.

Cheers

[1] *Director dynamic_dir returned no backend*
[2] *no backend connection*
[3] defaults to 60s in Varnish
[4] defaults to 5s in Apache HTTP Server

**Comment by markhowells** [comment]
Ah. That makes a lot of sense.  I did try changing some of the timeout settings - but I don;t think I went as low as 5s. I don't yet know what the backend server will ultimately be, so I just picked one I visit. 

I did wonder about whether it was a bug or not as I've was able to reproduce it so easily against a wide range of websites that the vmod would have had to be very broken.... 

I'll keep an eye on it, but feel free to close this issue if you're confident.

BTW - this is a _really_ useful vmod. We wouldn't be able to use Varnish without it as my use case doesn't necessarily know about all the backend servers in advance.

Thanks.
